File: ecopoint-microservice/services/booking-service/internal/models/booking.go
package models

import (
	"time"
)

// BookingStatus represents the status of a booking
type BookingStatus string

const (
	StatusPending     BookingStatus = "PENDING"
	StatusConfirmed   BookingStatus = "CONFIRMED"
	StatusAssigned    BookingStatus = "ASSIGNED"
	StatusInProgress  BookingStatus = "IN_PROGRESS"
	StatusCompleted   BookingStatus = "COMPLETED"
	StatusCancelled   BookingStatus = "CANCELLED"
	StatusFailed      BookingStatus = "FAILED"
)

// WasteType represents the type of waste
type WasteType string

const (
	WasteTypePlastic   WasteType = "PLASTIC"
	WasteTypePaper     WasteType = "PAPER"
	WasteTypeGlass     WasteType = "GLASS"
	WasteTypeMetal     WasteType = "METAL"
	WasteTypeOrganic   WasteType = "ORGANIC"
	WasteTypeElectronic WasteType = "ELECTRONIC"
	WasteTypeOther     WasteType = "OTHER"
)

// Address represents a physical address
type Address struct {
	Street    string  `json:"street" db:"street"`
	City      string  `json:"city" db:"city"`
	State     string  `json:"state" db:"state"`
	ZipCode   string  `json:"zip_code" db:"zip_code"`
	Latitude  float64 `json:"latitude" db:"latitude"`
	Longitude float64 `json:"longitude" db:"longitude"`
}

// Booking represents a waste collection booking
type Booking struct {
	ID             string       `json:"id" db:"id"`
	UserID         string       `json:"user_id" db:"user_id"`
	CollectorID    *string      `json:"collector_id" db:"collector_id"`
	WasteType      WasteType    `json:"waste_type" db:"waste_type"`
	Quantity       int          `json:"quantity" db:"quantity"`
	Description    string       `json:"description" db:"description"`
	PickupAddress  Address      `json:"pickup_address" db:"pickup_address"`
	ScheduledDate  time.Time    `json:"scheduled_date" db:"scheduled_date"`
	ScheduledTime  string       `json:"scheduled_time" db:"scheduled_time"`
	Status         BookingStatus `json:"status" db:"status"`
	EstimatedPrice float64      `json:"estimated_price" db:"estimated_price"`
	FinalPrice     *float64     `json:"final_price" db:"final_price"`
	Notes          string       `json:"notes" db:"notes"`
	Images         []string     `json:"images" db:"images"`
	CreatedAt      time.Time    `json:"created_at" db:"created_at"`
	UpdatedAt      time.Time    `json:"updated_at" db:"updated_at"`
}

// CreateBookingRequest represents the request to create a booking
type CreateBookingRequest struct {
	UserID         string    `json:"user_id" validate:"required"`
	WasteType      WasteType `json:"waste_type" validate:"required"`
	Quantity       int       `json:"quantity" validate:"required,min=1"`
	Description    string    `json:"description" validate:"required"`
	PickupAddress  Address   `json:"pickup_address" validate:"required"`
	ScheduledDate  string    `json:"scheduled_date" validate:"required"`
	ScheduledTime  string    `json:"scheduled_time" validate:"required"`
	Notes          string    `json:"notes"`
	Images         []string  `json:"images"`
}

// UpdateBookingRequest represents the request to update a booking
type UpdateBookingRequest struct {
	ID             string     `json:"id" validate:"required"`
	WasteType      *WasteType `json:"waste_type,omitempty"`
	Quantity       *int       `json:"quantity,omitempty"`
	Description    *string    `json:"description,omitempty"`
	PickupAddress  *Address   `json:"pickup_address,omitempty"`
	ScheduledDate  *string    `json:"scheduled_date,omitempty"`
	ScheduledTime  *string    `json:"scheduled_time,omitempty"`
	Notes          *string    `json:"notes,omitempty"`
	Images         []string   `json:"images,omitempty"`
}

// AssignCollectorRequest represents the request to assign a collector
type AssignCollectorRequest struct {
	BookingID   string `json:"booking_id" validate:"required"`
	CollectorID string `json:"collector_id" validate:"required"`
}

// UpdateStatusRequest represents the request to update booking status
type UpdateStatusRequest struct {
	BookingID string         `json:"booking_id" validate:"required"`
	Status    BookingStatus  `json:"status" validate:"required"`
	Notes     string         `json:"notes"`
}

// CancelBookingRequest represents the request to cancel a booking
type CancelBookingRequest struct {
	BookingID string `json:"booking_id" validate:"required"`
	Reason    string `json:"reason"`
}

// ListBookingsRequest represents the request to list bookings
type ListBookingsRequest struct {
	UserID      *string        `json:"user_id,omitempty"`
	CollectorID *string        `json:"collector_id,omitempty"`
	Status      *BookingStatus `json:"status,omitempty"`
	Page        int            `json:"page"`
	Limit       int            `json:"limit"`
}

// ListBookingsResponse represents the response for listing bookings
type ListBookingsResponse struct {
	Bookings []Booking `json:"bookings"`
	Total    int       `json:"total"`
	Page     int       `json:"page"`
	Limit    int       `json:"limit"`
}

File: ecopoint-microservice/mobile/flutter-app/lib/models/booking.dart
class Booking {
  final String id;
  final String userId;
  final String? collectorId;
  final WasteType wasteType;
  final double quantity;
  final String description;
  final Address pickupAddress;
  final DateTime scheduledDate;
  final String scheduledTime;
  final BookingStatus status;
  final double? estimatedPrice;
  final double? finalPrice;
  final String? notes;
  final List<String> images;
  final DateTime createdAt;
  final DateTime updatedAt;

  Booking({
    required this.id,
    required this.userId,
    this.collectorId,
    required this.wasteType,
    required this.quantity,
    required this.description,
    required this.pickupAddress,
    required this.scheduledDate,
    required this.scheduledTime,
    required this.status,
    this.estimatedPrice,
    this.finalPrice,
    this.notes,
    required this.images,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Booking.fromJson(Map<String, dynamic> json) {
    return Booking(
      id: json['id'] ?? '',
      userId: json['user_id'] ?? '',
      collectorId: json['collector_id'],
      wasteType: WasteType.fromString(json['waste_type'] ?? 'OTHER'),
      quantity: (json['quantity'] ?? 0).toDouble(),
      description: json['description'] ?? '',
      pickupAddress: Address.fromJson(json['pickup_address'] ?? {}),
      scheduledDate: DateTime.parse(json['scheduled_date'] ?? DateTime.now().toIso8601String()),
      scheduledTime: json['scheduled_time'] ?? '',
      status: BookingStatus.fromString(json['status'] ?? 'PENDING'),
      estimatedPrice: json['estimated_price']?.toDouble(),
      finalPrice: json['final_price']?.toDouble(),
      notes: json['notes'],
      images: List<String>.from(json['images'] ?? []),
      createdAt: DateTime.parse(json['created_at'] ?? DateTime.now().toIso8601String()),
      updatedAt: DateTime.parse(json['updated_at'] ?? DateTime.now().toIso8601String()),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'user_id': userId,
      'collector_id': collectorId,
      'waste_type': wasteType.value,
      'quantity': quantity,
      'description': description,
      'pickup_address': pickupAddress.toJson(),
      'scheduled_date': scheduledDate.toIso8601String().split('T')[0],
      'scheduled_time': scheduledTime,
      'status': status.value,
      'estimated_price': estimatedPrice,
      'final_price': finalPrice,
      'notes': notes,
      'images': images,
      'created_at': createdAt.toIso8601String(),
      'updated_at': updatedAt.toIso8601String(),
    };
  }
}

class Address {
  final String street;
  final String city;
  final String state;
  final String zipCode;
  final double latitude;
  final double longitude;

  Address({
    required this.street,
    required this.city,
    required this.state,
    required this.zipCode,
    required this.latitude,
    required this.longitude,
  });

  factory Address.fromJson(Map<String, dynamic> json) {
    return Address(
      street: json['street'] ?? '',
      city: json['city'] ?? '',
      state: json['state'] ?? '',
      zipCode: json['zip_code'] ?? '',
      latitude: (json['latitude'] ?? 0).toDouble(),
      longitude: (json['longitude'] ?? 0).toDouble(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'street': street,
      'city': city,
      'state': state,
      'zip_code': zipCode,
      'latitude': latitude,
      'longitude': longitude,
    };
  }

  String get fullAddress => '$street, $city, $state $zipCode';
}

enum WasteType {
  plastic('PLASTIC'),
  paper('PAPER'),
  glass('GLASS'),
  metal('METAL'),
  organic('ORGANIC'),
  electronic('ELECTRONIC'),
  other('OTHER');

  const WasteType(this.value);
  final String value;

  static WasteType fromString(String value) {
    switch (value.toUpperCase()) {
      case 'PLASTIC':
        return WasteType.plastic;
      case 'PAPER':
        return WasteType.paper;
      case 'GLASS':
        return WasteType.glass;
      case 'METAL':
        return WasteType.metal;
      case 'ORGANIC':
        return WasteType.organic;
      case 'ELECTRONIC':
        return WasteType.electronic;
      default:
        return WasteType.other;
    }
  }
}

enum BookingStatus {
  pending('PENDING'),
  confirmed('CONFIRMED'),
  assigned('ASSIGNED'),
  inProgress('IN_PROGRESS'),
  completed('COMPLETED'),
  cancelled('CANCELLED'),
  failed('FAILED');

  const BookingStatus(this.value);
  final String value;

  static BookingStatus fromString(String value) {
    switch (value.toUpperCase()) {
      case 'PENDING':
        return BookingStatus.pending;
      case 'CONFIRMED':
        return BookingStatus.confirmed;
      case 'ASSIGNED':
        return BookingStatus.assigned;
      case 'IN_PROGRESS':
        return BookingStatus.inProgress;
      case 'COMPLETED':
        return BookingStatus.completed;
      case 'CANCELLED':
        return BookingStatus.cancelled;
      case 'FAILED':
        return BookingStatus.failed;
      default:
        return BookingStatus.pending;
    }
  }
}